Rework Pretrained Tokenizer and Indexer

maksym-del opened the pull request:
The idea is to both tokenize and index tokens in the tokenizer while adding token ids to the Token data structure. In this setup, PretrainedTokenIndexer only needs to pick up correct token id as well as create allennlp vocabulary from the pretrained one.
 Thus we subclass  SingleIdTokenIndexer. 
To get special tokens handled, we first index the string with huggingface method that adds special token ids as well and then converting indexes back to string tokens (to be used by AllenNLP tokens).
This feels to me the cleanest overall approach.

We also now have token_type_id field added to the token. It can now be computed automatically with huggingface's code. I do not use it in any places yet since it is not used by recent transformers like XLM or RoBERTa. However, future PR can better utilize it.

Fixes https://github.com/allenai/allennlp/issues/3356 and answers the question in https://github.com/allenai/allennlp/issues/3224.

maksym-del commented:
This one is ready to be reviewed.

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
     We take a model name as an input parameter, which we will pass to
     ``AutoTokenizer.from_pretrained``.
 
+    We also add special tokens relative to the pretrained model and trunkate the sequences.
```
`s/trunkate/truncate/`

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
 
+    We also add special tokens relative to the pretrained model and trunkate the sequences.
+
+    This tokenizer also indexes tokens and adds the indexes as the ``Token`` fields so that
```
`s/as the/to the/`, and a period at the end of the sentence.

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
-        If given, these tokens will be added to the beginning of every string we tokenize.  We try
-        to be a little bit smart about defaults here - e.g., if your model name contains ``bert``,
-        we by default add ``[CLS]`` at the beginning and ``[SEP]`` at the end.
+        If given, these tokens will be added to the beginning of every string we tokenize.
```
I'd probably just remove these parameters, as `add_special_tokens` covers the intended use for `start_tokens` and `end_tokens`.  If we really want to keep them, we need to note that these get added _before_ adding special tokens.

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
+        token_ids = encoded_tokens["input_ids"]
+        token_type_ids = encoded_tokens["token_type_ids"]
 
+        token_strings = self._tokenizer.convert_ids_to_tokens(token_ids, skip_special_tokens=False)
```
Seems like this is an extra iteration over the tokens that we don't need, given the iteration you're doing below.  Why not remove this line and do a simple vocab lookup inside the loop?

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
````diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
         # ``transformers``` is dealing with the whitespace...
-        token_strings = self._start_tokens + self._tokenizer.tokenize(text) + self._end_tokens
-        return [Token(t) for t in token_strings]
+        if ALLENNLP_SENTENCE_PAIR_SEP in text:
````
I'd be happier if we added a `tokenize_sentences(sentences: List[str]) -> List[Token]` method to `Tokenizer`, instead of adding this special constant.  But I suppose this could do for now...  I'm just wary of adding a recommended public constant that we might want to remove later.

maksym-del replied:
Yeah, but the downside of it is that we cannot handle a sentence pair with an inherited ``tokenize`` method in this setup. So I introduced this symbol to tell this method that it gets a sentence pair. Optional argument to the ``tokenize`` might be a cleaner solution, but this breaks API. 

I am not sure however how important it is to handle a sentence pair in the main tokenize method and I followed your suggestion.

Regarding 3+ sentences, to the best of my knowledge, it will be a custom logic that is not available in transformers repo. Probably for another PR.

matt-gardner replied:
Well, either way the `DatasetReader` has to do something differently to handle this.  In your version, you have to manually add the separator token, then have the tokenizer immediately undo this concatenation and process things accordingly.  This also means that we'd have to modify *every* tokenizer to handle this token, because the `DatasetReader` would be adding it, and our other tokenizers would see it if you switched.

In my way, the default behavior is just to concatenate the sentences (probably with a space) and pass it to `tokenize()`.  No other tokenizers break, and so the `DatasetReader` doesn't have to have custom logic that's aware of which `tokenizer` it's using.

maksym-del replied:
I see your point regarding the special token, it totally makes sense to omit it since we might want to change the tokenizer type we are using.

Regarding concatenating sentence pair with space, how will you then decide where to split it in ``tokenize``? Huggingface's ``encode_plus`` method expects sentences to come separated.

matt-gardner replied:
I was talking about the default implementation in `Tokenizer.tokenize_sentences`, which should be easy.  `PretrainedTransformerTokenizer` would override `tokenize_sentences` with the logic you have here.

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -9,6 +9,8 @@
 
 logger = logging.getLogger(__name__)
 
+ALLENNLP_SENTENCE_PAIR_SEP = "@@SEP_SENTPAIR@@"
```
As I said in another comment, I'm not thrilled with having this.  But if we do want it, it should not belong here, it should be in something like `data/common.py`.  And I would name it `SENTENCE_SEPARATOR`, instead of what you have.

matt-gardner commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -109,13 +82,26 @@ def tokenize(self, text: str) -> List[Token]:
+
         return tokens
+
+    def tokenize_sentence_pair(self, sentence_1: str, sentence_2: str):
```
I was suggesting putting this on the base class, but I suppose putting it here for now is fine, and we can add it higher up later if necessary.  This just means that the dataset reader has to instantiate a `PretrainedTransformerTokenizer`, instead of just a `Tokenizer`, for the types to work out.

maksym-del replied:
I added the commit representing this approach.

DeNeutoy commented on allennlp/data/tokenizers/tokenizer.py:
```diff
@@ -47,6 +47,18 @@ def tokenize(self, text: str) -> List[Token]:
         """
         raise NotImplementedError
 
+    def tokenize_sentences(self, sentences: List[str]) -> List[Token]:
```
Why does this need to be on the base class? Seems like it should be a private method on the `pretrained_transformer_tokenizer` class. It has confusing behaviour that is only desirable for tokenising text for a language model.

maksym-del replied:
If it is a private method and it is used somewhere in DatasetReader, it will mean that this DatasetReader will only work with this type of tokenizer.

maksym-del replied:
However, I am not against making it private again. It will still be useful in many use cases,
and as @matt-gardner pointed out, we can always add it higher up later if necessary.

DeNeutoy replied:
Yes - my point is simply that currently, if you change the tokenizer to one which is not a pretrained lm tokenizer, what happens is pretty undesirable (your two sentences are tokenised as though they are a single sentence). Is this something you can handle in a `DatasetReader` without detecting that you aren't using one of the LM specific tokenizers in the first place? 

We can wait for Matt's opinion, but i'd suggest removing this.

maksym-del replied:
@DeNeutoy do you mean removing it completely or just making it private?

matt-gardner replied:
We talked about this internally at a review meeting today, and I agree with @DeNeutoy's strong counterexample about the generality of this approach.  You can't just change tokenizers to go from BERT-style encoding to something else, you'd have to add new `TextFields` and such.  So it doesn't really make sense to have the method on the super class as I recommended.  Sorry.  We should move it back to a (public) method just on `PretrainedTransformerTokenizer`.  And given that, we probably don't need to handle more than two sentences, and just giving a sentence pair method for now is fine.  Sorry for the extra work that I recommended.

maksym-del replied:
Done!

DeNeutoy commented on allennlp/data/tokenizers/pretrained_transformer_tokenizer.py:
```diff
@@ -22,36 +22,101 @@ class PretrainedTransformerTokenizer(Tokenizer):
+        return tokens
+
+    @overrides
+    def tokenize_sentences(self, sentences: List[str]) -> List[Token]:
```
This seems to be too general, even as a method on this specific class - it's not possible to tokenise more than 2 sentences, so why allow it to be passed more than two sentences?

maksym-del replied:
I thought in the future we could implement automatic addition of special tokens in case of 3+ sentences. This might be useful for BERT based reading comprehension or summarization.

DeNeutoy commented on allennlp/tests/data/token_indexers/pretrained_transformer_indexer_test.py:
```diff
@@ -3,24 +3,74 @@
-    def test_as_array_produces_token_sequence(self):
+    def test_as_array_produces_token_sequence_bert(self):
         tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+        allennlp_tokenier = PretrainedTransformerTokenizer("bert-base-uncased")
```
allennlp_tokenier should be `allennlp_tokenizer`

brendan-ai2 commented on allennlp/tests/data/token_indexers/pretrained_transformer_indexer_test.py:
```diff
@@ -3,24 +3,74 @@
         tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+        allennlp_tokenier = PretrainedTransformerTokenizer("bert-base-uncased")
+
+        # uncased
```
A block level comment like this suggests that the block should be broken out into its own test case. (Especially since the blocks don't appear to rely on much shared state.)

maksym-del replied:
Done

brendan-ai2 commented on allennlp/tests/data/token_indexers/pretrained_transformer_indexer_test.py:
```diff
@@ -3,24 +3,74 @@
+        expected_ids = tokenizer.convert_tokens_to_ids(tokens)
+        # tokens tokenized with our pretrained tokenizer have indices in them
+        allennlp_tokens = allennlp_tokenier.tokenize(string_no_specials)
+        vocab = Vocabulary()
```
I think we might want some expectations on the vocabulary produced. (In a separate test case.) Nothing too heavyweight. Perhaps:

1. Sanity check that just verifies that a short simple sentence produces an expected vocabulary.
2. If it's possible to have tokens that aren't indexed by the tokenizer, introduce some. Verify that the indices of the other tokens don't change.

maksym-del replied:
Thanks @brendan-ai2 ! Your first sanity check made me discover https://github.com/allenai/allennlp/issues/3456

brendan-ai2 commented on allennlp/tests/data/tokenizers/pretrained_transformer_tokenizer_test.py:
```diff
@@ -17,11 +50,32 @@ def test_splits_cased(self):
         assert tokens == expected_tokens
 
-    def test_splits_uncased(self):
+        # sentence pair
```
Similarly, this would be cleaner as a separate test case.

maksym-del replied:
Done

brendan-ai2 commented on allennlp/data/token_indexers/pretrained_transformer_indexer.py:
```diff
@@ -9,16 +9,17 @@
     """
-    This :class:`TokenIndexer` uses a tokenizer from the ``transformers`` repository to
-    index tokens.  This ``Indexer`` is only really appropriate to use if you've also used a
+    This ``TokenIndexer`` assumes that Tokens already have their indexes in them.
```
Is it possible to assert the indices are present in the tokens already? If not, see my comment about a test to distinguish between tokens that already have indices and those that don't.

maksym-del replied:
Added a check to the indexer

maksym-del replied:
Adding this check also spoiled potential bugs in masked_lm and casual_lm readers.

maksym-del commented:
I addressed feedback and also moved BERT vocab creation inside the ```Vocabulary``` constructor. It fixes https://github.com/allenai/allennlp/issues/3097.

There are 2 tests failing that I do not see to be related to the code I changed... Also, all my local pytest checks pass.

So I believe this PR is ready for review.

matt-gardner commented on allennlp/data/vocabulary.py:
```diff
@@ -662,6 +681,32 @@ def extend_from_instances(
             min_pretrained_embeddings=min_pretrained_embeddings,
         )
 
+    def _extend_from_transformers_vocabulary(
```
Can you move this back inside the token indexer call?  I don't think this is the right fix to #3097.  I'd like to avoid extra parameters that you have to remember to set in the config file; it just makes everything harder to use.  I think there's a different solution that we can come up with, probably just in the timing of when we save the vocab in the trainer code.

maksym-del replied:
Done

maksym-del commented on allennlp/data/dataset_readers/next_token_lm.py:
```diff
@@ -78,5 +85,7 @@ def text_to_instance(
         fields: Dict[str, Field] = {"tokens": input_field}
         if target:
-            fields["target_ids"] = TextField([Token(target)], self._token_indexers)
+            wordpiece = self._targets_tokenizer.tokenize(target)[0]
```
Done

maksym-del commented:
I addressed the feedback. All the relevant tests pass!

maksym-del commented:
Thanks!
